internal_packages/keybase/lib/pgp-key-store.cjsx

```diff
@@ -0,0 +1,391 @@
+NylasStore = require 'nylas-store'
+{DraftStore, MessageBodyProcessor, RegExpUtils} = require 'nylas-exports'
+{remote} = require 'electron'
+kb = require './keybase'
+pgp = require 'kbpgp'
+_ = require 'underscore'
+path = require 'path'
+fs = require 'fs'
+
+class PGPKeyStore extends NylasStore
+
+  constructor: ->
+    super()
+
+    @_pubKeys = {}
+    @_privKeys = []
+
+    @_msgCache = []
+    @_msgStatus = []
+
+    @_pubWatcher = null
+    @_privWatcher = null
+
+    @_keyDir = path.join(NylasEnv.getConfigDirPath(), 'keys')
+    @_pubKeyDir = path.join(@_keyDir, 'public')
+    @_privKeyDir = path.join(@_keyDir, 'private')
+
+    # Create the key storage file system if it doesn't already exist
+    fs.access(@_keyDir, fs.R_OK | fs.W_OK, (err) =>
+      if err
+        fs.mkdir(@_keyDir, (err) =>
+          if err
+            console.warn err
+          else
+            fs.mkdir(@_pubKeyDir, (err) =>
+              if err
+                console.warn err
+              else
+                fs.mkdir(@_privKeyDir, (err) =>
+                  if err
+                    console.warn err
+                  else
+                    @watch())))
+      else
+        fs.access(@_pubKeyDir, fs.R_OK | fs.W_OK, (err) =>
+          if err
+            fs.mkdir(@_pubKeyDir, (err) =>
+              if err
+                console.warn err))
+        fs.access(@_privKeyDir, fs.R_OK | fs.W_OK, (err) =>
+          if err
+            fs.mkdir(@_privKeyDir, (err) =>
+              if err
+                console.warn err))
+        @_populate(isPub = true)
+        @_populate(isPub = false)
+        @watch())
+
+    @unlisten = DraftStore.listen(@_onDraftChanged, @)
+
+  validAddress: (address, isPub) =>
+    if (!address || address.length == 0)
+      @_displayError('You must provide an email address.')
+      return false
+    if not (RegExpUtils.emailRegex().test(address))
+      @_displayError('Invalid email address.')
+      return false
+    keys = if isPub then @pubKeys(address) else @privKeys(address, timed=false)
+    keystate = if isPub then 'public' else 'private'
+    if (keys.length > 0)
+      @_displayError("A PGP #{keystate} key for that email address already exists.")
+      return false
+    return true
+
+  ### I/O and File Tracking ###
+
+  watch: () =>
+    if (!@_pubWatcher)
+      @_pubWatcher = fs.watch(@_pubKeyDir, () => @_populate(isPub = true))
+    if (!@_privWatcher)
+      @_privWatcher = fs.watch(@_privKeyDir, () => @_populate(isPub = false))
+
+  unwatch: () =>
+    if (@_pubWatcher)
+      @_pubWatcher.close()
+    @_pubWatcher = null
+    if (@_privWatcher)
+      @_privWatcher.close()
+    @_privWatcher = null
+
+  _populate: (isPub) =>
+    # add metadata elements (sans keys) to later be populated with actual keys
+    # from disk
+    if isPub
+      keyDirectory = @_pubKeyDir
+      @_pubKeys = []
+    else
+      keyDirectory = @_privKeyDir
+      @_privKeys = []
+    fs.readdir(keyDirectory, (err, filenames) =>
```

Comment:
It' be really nice to consider making a class to represent these "key file" hashes ({path, addresses}) so that we could change the implementation totally within that class in the future. I imagine this could look something like this:

```
fs.readdir(keyDirectory, (err, filenames) =>
   filenames.filter((name) => name isnt '.').map((name) => {
       new SavedKey(path.join(keyDirectory, filename))
   })


```